content/blog/2019-10-02-codestitcher.md

```diff
@@ -0,0 +1,166 @@
+
+There *is* is a 1/2-approximation algorithm for this problem guarantees that the weight of the path cover is within *1/2* of the optimal solution. While its performance does not absolutely exceed that of PH's heuristic, Codestitcher uses a hybrid of PH's method and the approximation algorithm, which provides better results than either algorithm run individually.
+
+### Hierarchical Code Collocation
+
+Coming up with a framework to integrate intraprocedural and interprocedural ordering optimizations while taking *layout distance* (the distance between two instructions) into account is difficult. We will first try to formalize this problem.
+
+#### *d*-Close Partial Layout
+
+It would help if we could formally define what layout distance really means. The paper's definition is quite intuitive: the number of bytes between two basic blocks. To take advantage of spatial locality, we want to ensure that related instructions that are executed frequently have as few bytes between them as possible. (This is because the instruction cache can only hold so many instructions.)
+
+Given a specific distance value *d*, we can try to lay out our program to maximize the number of times control is transferred between two basic blocks within *d* bytes of each other. However, this is not enough. We need to optimize over many different *d* values to account for hierarchical memory systems with multiple cache levels. Our goal is to find the *finest grain* layout that maximizes control transfers across all combinations of basic blocks, across multiple distances.
+
+How do we solve this? The paper suggests a greedy approach that encodes the problem as a graph of basic block sequences, with edges weighted by the number of control flow transfers within *d* bytes. Nodes of the graph are iteratively coalesced in order of descending weight until no edges remain, and the remaining node's ordering is output.
+
+### Finishing the Codestitcher layout optimization
+
+Using the *d*-close partial layout results, Codestitcher orders the final basic block chains based on a ratio of their total execution counts and size in bytes. This layout is then used to recompile an optimized version of the original program.
```

Comment:
> a ratio of their total execution counts and size in bytes.

This is pretty surprising—can you explain a little more about why that ratio is expected to be meaningful?